crates/reward/src/lib.rs

```diff
@@ -0,0 +1,269 @@
+    }
+
+    fn balance_to_fixed(x: T::Balance) -> Result<SignedFixedPoint<T, I>, DispatchError> {
+        let unsigned_fixed_point = SignedFixedPoint::<T, I>::checked_from_integer(Self::balance_to_inner(x)?)
+            .ok_or(Error::<T, I>::TryIntoIntError)?;
+        Ok(unsigned_fixed_point)
+    }
+
+    pub fn participants() -> Vec<(T::AccountId, T::Balance)> {
+        <Stake<T, I>>::iter().collect()
+    }
+}
+
+pub trait Rewards<AccountId> {
+    /// The balance type for a currency.
+    type Balance: AtLeast32BitUnsigned + FullCodec + Copy + MaybeSerializeDeserialize + Debug + Default;
+
+    fn get_stake(account_id: &AccountId) -> Self::Balance;
+
+    fn deposit_stake(account_id: &AccountId, amount: Self::Balance) -> Result<(), DispatchError>;
+
+    fn distribute(reward: Self::Balance) -> Result<Self::Balance, DispatchError>;
+
+    fn compute_reward(account_id: &AccountId) -> Result<Self::Balance, DispatchError>;
+
+    fn withdraw_stake(account_id: &AccountId, amount: Self::Balance) -> Result<(), DispatchError>;
+
+    fn withdraw_reward(account_id: &AccountId) -> Result<Self::Balance, DispatchError>;
+}
+
+impl<T: Config<I>, I: 'static> Rewards<T::AccountId> for Pallet<T, I>
+where
+    T::Balance: MaybeSerializeDeserialize + Debug,
+{
+    type Balance = T::Balance;
+
+    fn get_stake(account_id: &T::AccountId) -> Self::Balance {
+        Self::stake(account_id)
+    }
+
+    fn deposit_stake(account_id: &T::AccountId, amount: Self::Balance) -> Result<(), DispatchError> {
+        <TotalStake<T, I>>::mutate(|total_stake| {
+            *total_stake = total_stake
+                .checked_add(&amount)
+                .ok_or(Error::<T, I>::ArithmeticOverflow)?;
+            Ok::<_, Error<T, I>>(())
+        })?;
+
+        <Stake<T, I>>::mutate(account_id, |stake| {
+            *stake = stake.checked_add(&amount).ok_or(Error::<T, I>::ArithmeticOverflow)?;
+            Ok::<_, Error<T, I>>(())
+        })?;
+
+        let amount_as_fixed = Self::balance_to_fixed(amount)?;
+        <RewardTally<T, I>>::mutate(account_id, |reward_tally| {
+            let reward_per_token = Self::reward_per_token();
+            let reward_per_token_mul_amount = reward_per_token
+                .checked_mul(&amount_as_fixed)
+                .ok_or(Error::<T, I>::ArithmeticOverflow)?;
+            *reward_tally = reward_tally
+                .checked_add(&reward_per_token_mul_amount)
+                .ok_or(Error::<T, I>::ArithmeticOverflow)?;
+            Ok::<_, Error<T, I>>(())
+        })?;
+
+        Ok(())
+    }
+
+    fn distribute(reward: Self::Balance) -> Result<Self::Balance, DispatchError> {
+        let total_stake = Self::total_stake();
+        if total_stake.is_zero() {
+            return Ok(Self::Balance::zero());
+        }
+
+        let reward_per_token = Self::reward_per_token();
+        let reward_as_fixed = Self::balance_to_fixed(reward)?;
+        let total_stake_as_fixed = Self::balance_to_fixed(total_stake)?;
```

Comment:
Not sure what the solution is here. This is probably fine for relayer `stake` if that is just the SLA (0-100) but the concern would be for unbounded vault `stake`. To minimise the chance this would overflow we could set `stake` here directly to the vault's `issued_tokens` which we probably care more about - otherwise anyone can get more rewards for locking more `backing_collateral` and preventing issues by setting `active = false`.